### PR TITLE
Plugin Support

### PR DESCRIPTION
**Pull request by BillSchumacher**
### Background

Adds basic plugin support.

### Changes

Add plugins.py which handles scanning plugins folder for zipfiles and check for modules to import.

In main.py on startup it will load the plugins into the config file.

In LLM utils it will iterate the plugins on each response.

### Documentation

Docstrings and README updated.

### Test Plan

I tested this code here: https://github.com/BillSchumacher/Auto-Vicuna

I have an example plugin for that project here: https://github.com/BillSchumacher/Auto-Vicuna-Its-Alive

It would probably work, but it allows optional return in on_response.

However, I did make changes like add a whitelist / blacklist as discussed.

There's a chance I broke something. This was mostly a copy/paste.

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thouroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes 

I have not tested this at all, I hope a lot of people make plugins and do test.

### CODE REVIEW

**Comment by Jordan-Mesches** [comment]
This is pretty similar to #350

**Comment by BillSchumacher** [comment]
> This is pretty similar to #350

Not really this loads external repos as zipfiles.

**Comment by BillSchumacher** [comment]
If anyone has suggestions on other ways to hook certain points in the Auto-GPT flow don't hestitate to speak up.

**Comment by Torantulino** [comment]
> If anyone has suggestions on other ways to hook certain points in the Auto-GPT flow don't hesitate to speak up.

This is a key point we need to work out, lots of different ways to do this, and we want to focus on enabling as many types of plugins as possible!

File: README.md
Comment (Torantulino) [inline README.md]: I think this needs to be further stressed.
As you've said, there's plenty of room for malicious plugins.

**Comment by Jordan-Mesches** [comment]
> Not really this loads external repos as zipfiles.

I'm just curious as to how plugins would function differently than commands.

**Comment by Torantulino** [comment]
Plugins can implement new commands.
The difference is that they will allow Auto-GPT to be hooked up to _anything_.

There are plenty of use-cases that people have requested that don't make sense being in the core repo for everyone.

**Comment by BillSchumacher** [comment]
> > Not really this loads external repos as zipfiles.
> 
> I'm just curious as to how plugins would function differently than commands.

It would keep this codebase clean, allowing for faster reviews and more flexibility for people to add functionality they think other people would like without having to wait for a merge.

**Comment by Jordan-Mesches** [comment]
Ah, yeah I get it. I didn't really consider bloat being a possible concern, but it's a valid point.

**Comment by Jordan-Mesches** [comment]
Would some kind of "plugin marketplace" be needed to support this?

**Comment by BillSchumacher** [comment]
> Would some kind of "plugin marketplace" be needed to support this?

The current implementation allows for any repo that uses the template to be downloaded as zip and dropped into the plugins folder.

I would imagine a plugins channel in discord to start with.

Toran might get rich with a 30% cut from the marketplace though 🤔

File: README.md
Comment (BillSchumacher) [reply README.md]: ✅

**Comment by BillSchumacher** [comment]
> Just thinking through the plugin whitelist and possibility for any abuse. It only checks by name. I think the caveat emptor in the readme is probably the best we'll get. Possible improvement - we're only checking by name, and we don't check if we've already loaded a whitelisted plugin by that name - which might be worth adding?

That's a fair point this implementation is actually quite bad.

We would probably want some md5 hashes or sha256 of the zip.

I also didn't take into account external dependencies, which might be a good thing but probably to result in complaints.

**Review by nponeccop — DISMISSED** [review]

**Comment by mikekelly** [comment]
@richbeales @BillSchumacher is there a way to do it where they're added as python packages in requirements.txt? The README could recommend specifying the specific commit like this:

```
git+https://github.com/random-dev/example-auto-gpt-plugin@42b95dc=example-auto-gpt-plugin
```

Whatever package delivery option you pick; I'd recommend, to keep the feature simple for now, that users have to manually authorise each plugin every time the application boots, unless an environment variable `SKIP_PLUGIN_AUTH=true`. If that env var is set then the UI should display a clear warning and list the plugins that are registered.

**Comment by TaylorBeeston** [comment]
@BillSchumacher does this seem like a reasonable API to land on? 
image

```python
PromptName = str
PromptRole = str
PromptGoals = List[str]
Prompt = tuple[PromptName, PromptRole, PromptGoals]

post_prompt(prompt: Prompt) -> Prompt

on_planning(prompt: Prompt, messages: List[str]) -> Optional[str]
post_planning(response: str) -> str

pre_instruction(messages: List[str]) -> List[str]
on_instruction(messages: List[str]) -> Optional[str]
# TODO: Better JSON type
post_instruction(response: str) -> str

pre_command(instruction: str) -> str
post_command(messages: List[str]) -> List[str]
```

**Comment by TaylorBeeston** [comment]
Where every hook is optional, and it `on_planning`/`on_instruction` allow you to leave the default OpenAI calls by simply returning `None`

**Comment by TaylorBeeston** [comment]
As an example for this kind of architecture, the guy on discord who swapped out JSON instructions with YAML could override `pre_instruction` to ask for YAML instead of JSON, then override `pre_command` to convert the YAML back to json so that existing commands still work unchanged

**Comment by TaylorBeeston** [comment]
Also allows adding more commands via a `custom_commands` field that you can throw in

**Comment by TaylorBeeston** [comment]
Btw, I have no idea how Figma deals with people collaborating for free in FigJam files, but in case it's allowed, here's the link to that FigJam: https://www.figma.com/file/KpX21VXXmyXDZhTOfBH1sL/AutoGPT-Plugin-Architecture?node-id=0%3A1&t=vzGe8FlrCbMMvl4R-1

**Comment by BillSchumacher** [comment]
> @BillSchumacher does this seem like a reasonable API to land on? image
> 
> ```python
> PromptName = str
> PromptRole = str
> PromptGoals = List[str]
> Prompt = tuple[PromptName, PromptRole, PromptGoals]
> 
> post_prompt(prompt: Prompt) -> Prompt
> 
> on_planning(prompt: Prompt, messages: List[str]) -> Optional[str]
> post_planning(response: str) -> str
> 
> pre_instruction(messages: List[str]) -> List[str]
> on_instruction(messages: List[str]) -> Optional[str]
> # TODO: Better JSON type
> post_instruction(response: str) -> str
> 
> pre_command(instruction: str) -> str
> post_command(messages: List[str]) -> List[str]
> ```

Yeah I like it =)

**Comment by BillSchumacher** [comment]
I'll be implementing this shortly.

**Comment by BillSchumacher** [comment]
Implementation note, instead of the prompt tuple I'm passing a PromptGenerator object, which is what we use to contruct the full system prompt. 

If you change something in post_prompt, it will alter the system_prompt. Something we've been thinking about doing with the assistants.

**Comment by BillSchumacher** [comment]
Another note, when creating an AI config you can now pass a `PromptGenerator` object.

The prompt generator now also has a goals property that is assigned from the `AIConfig` object when construct_full_prompt is called.

**Comment by BillSchumacher** [comment]
I guess I need to move the name and role into the prompt generator as well...

**Comment by TaylorBeeston** [comment]
@BillSchumacher no worries! Those were just types I'd thought of that might make sense from my black box view. I will update the Figma diagram! I'm going to work on some documentation for all of this as you work on implementation to help the community write/read plug-ins, so please lmk if anything else needs to change like a different name or anything like that!

**Comment by BillSchumacher** [comment]
Implementation note, even though we are passing a PromptGenerator object to the on_planning method of the plugin changing it's values will not change the prompt. It should be used to come up with additional context to send to the chat completion function.

```
on_planning(prompt: PromptGenerator, messages: List[str]) -> Optional[str]
```

If the response is None or an empty string it will be skipped, otherwise it will be added to the context **if** there are enough send tokens available, if there are not the plugin processing loop stops early and the context is not added.

**Comment by BillSchumacher** [comment]
I'm not sure if the order of the context matters for getting the correct result back.

**Comment by BillSchumacher** [comment]
It's also noteworthy that if you wanted, since this is python you could technically alter the context and that updated context would remain, if you wanted to remove certain messages, add new ones or whatever. Kind of a hack but it would work.

**Comment by BillSchumacher** [comment]
Our workflow doesn't quite work like this so I'm going to try to update the diagram, the sub-agents are not called every loop but rather by command.

**Comment by BillSchumacher** [comment]
image

**Comment by BillSchumacher** [comment]
Alright, so `pre_instruction` and `on_instruction` are called when the `start_agent` or `message_agent` commands are executed.

`pre_instruction` recieves a list of messages which is actually the single user message and the result should be a list of messages, which is then extended onto the messages it was passed as messages from "system".

`on_instruction` recieves the full context of the resulting chat completion from the assistant and that is added to the assistant's context.

`post_instruction` takes the reply from the assistant and either passes the original or modifies it.

Might want to change some of the names here, I'm not sure they make sense.

**Comment by BillSchumacher** [comment]
or we might implement it differently, definitely would like some review.

**Comment by SlistInc** [comment]
Just a quick thought: it may be necessary to assign some "capability" tag to a plugin so as to allow other plugins to require that "capability". For example a fiverr integration plugin with the ability to hire freelancer on it's own may require another plugin with a "payment" capability and a "budget" call.

Sorry if this is just a thought and not fleshed out, it may make sense to keep it in mind.

**Comment by BillSchumacher** [comment]
> Just a quick thought: it may be necessary to assign some "capability" tag to a plugin so as to allow other plugins to require that "capability". For example a fiverr integration plugin with the ability to hire freelancer on it's own may require another plugin with a "payment" capability and a "budget" call.
> 
> Sorry if this is just a thought and not fleshed out, it may make sense to keep it in mind.

Yeah we definitely need more input like that. Certainly not fleshed out good.

**Comment by BillSchumacher** [comment]
`post_planning` takes the reply from the `on_planning` result and can either pass the original value or modify it, this is the dreaded JSON response.

**Comment by BillSchumacher** [comment]
`post_planning` would be a great spot for someone to make a plugin that deals with bad output appropriately ;)

**Comment by BillSchumacher** [comment]
`pre_command` takes in `command_name`: str and `arguments`: Dict[str, Any] and the plugin can pass the original values or modify them.
`post_command` takes in `command_name` and `result`: str of the executed command and either returns the original `result` or modifies it.

**Comment by BillSchumacher** [comment]
Regarding custom commands, in the `post_prompt` hook when you recieve the `PromptGenerator` object you can call the function 

```
def add_command(
        self,
        command_label: str,
        command_name: str,
        args=None,
        function: Optional[Callable] = None,
    )
```

earlier I added a new arg called function, which takes any callable object, thus you can add custom functions.

**Comment by BillSchumacher** [comment]
I'll add the implementation to get and execute command soon.

**Comment by SlistInc** [comment]
> `post_planning` would be a great spot for someone to make a plugin that deals with bad output appropriately ;)

I think it may make sense that in some steps (e.g. like this one) the plugins see which other plugin are active/available in order to change behaviour? I might want to react to the fact that there is a vicuna model plugin active and deal with outputs differently?

It may introduce some race conditions though and plugins should never assume their order of execution.

**Comment by BillSchumacher** [comment]
Yeah the ordering is going to be left up to the user. we could on plugin initialization pass the config object to the plugin.

**Comment by BillSchumacher** [comment]
Actually it isn't I lied. They will be in alphabetical order, by default.

**Comment by BillSchumacher** [comment]
Anyone that's interested in contributed to this before we merge it can checkout my plugin-support branch and submit PRs.

**Comment by TaylorBeeston** [comment]
@BillSchumacher @SlistInc ordering could get done by prefixing plugin names with numbers e.g. 10-Vicuna.zip 20-Output-Fixer.zip

**Comment by BillSchumacher** [comment]
> @BillSchumacher @SlistInc ordering could get done by prefixing plugin names with numbers e.g. 10-Vicuna.zip 20-Output-Fixer.zip

Yeah that's a good point.

**Comment by BillSchumacher** [comment]
I'm going to update the template repo now.

**Comment by TaylorBeeston** [comment]
Also, I have a feeling this will evolve into packages up on pypi that a wrapper then pieces together rather than loading zip files directly in this repo, so order could be much more tightly controlled that way

**Comment by TaylorBeeston** [comment]
E.g.
```python
from autogpt import init
from autogpt-vicuna import vicuna_plugin
from autogpt-yaml import yaml_plugin

init(plugins=[vicuna_plugin, yaml_plugin]) 
```

**Comment by BillSchumacher** [comment]
> Also, I have a feeling this will evolve into packages up on pypi that a wrapper then pieces together rather than loading zip files directly in this repo, so order could be much more tightly controlled that way

Yeah I just took the requirements and made it work =) I don't care how they get loaded at the end of the day.

**Comment by BillSchumacher** [comment]
Please submit PRs to my [plugin-](https://github.com/BillSchumacher/Auto-GPT/tree/plugin-support) branch if you want to see this look different. I don't think this code is great or final.

**Comment by TaylorBeeston** [comment]
No worries! I'm just looking at this through the lens of what I'd do in js-land as my python is not as good 😅

**Comment by BillSchumacher** [comment]
Alright, template is updated as well, let me know if I messed something up.

**Comment by richbeales** [comment]
LGTM

**Comment by nponeccop** [comment]
@BillSchumacher There are conflicts and red CI now (I know you know it)

**Comment by TaylorBeeston** [comment]
@BillSchumacher thoughts on turning this into a monorepo so that we can reuse the same ABC across plugin template repo and the main repo? In TS land I would just make this a pnpm workspace, but I'm not sure if there's an equivalent for python

**Comment by TaylorBeeston** [comment]
I can do some searching and set that up if it seems worth it, but don't want to waste time if it isn't warranted

**Comment by BillSchumacher** [comment]
> I can do some searching and set that up if it seems worth it, but don't want to waste time if it isn't warranted

It's definitely worth discussing before doing, if the community thinks we should go one direction or the other please speak up.

**Comment by BillSchumacher** [comment]
We now have the functionality of the command registry, as soon as I get it all cleaned up.

**Comment by nponeccop** [comment]
@BillSchumacher Conflicts ;)

**Comment by evahteev** [comment]
> Please submit PRs to my [plugin-support](https://github.com/BillSchumacher/Auto-GPT/tree/plugin-support) branch if you want to see this look different. I don't think this code is great or final.

Thanks for huge work to move it forward!. I'm currently working on compatibility with OpenAi plugins so we would be able to utilize the manifest open API standard, and would be able to hook up to the existing infrastructure of OpenAI plugins. I believe I'm pretty close to finishing, so submitting PR to your brunch shortly.

**Comment by keeganmccallum** [comment]
something like this would be sweet to allow for unlimted plugins: https://github.com/microsoft/visual-chatgpt/tree/main/TaskMatrix.AI

**Comment by BillSchumacher** [comment]
I've added a new hook ```handle_chat_completion``` which takes `message`, `model`, `temperature` and `max_tokens`

I also added functions to abort early if the various hooks were not enabled by the plugin.

**Comment by SlistInc** [comment]
Some observations:

- By mistake I created the "plugins" folder into "Auto-GPT/plugins/" rather than "Auto-GPT/autogpt/plugins/"
- The absence of the plugins folder caused this error on my system (macos ventura 13.3, M2):
`Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "blabla/Auto-GPT/autogpt/__main__.py", line 97, in <module>
    main()
  File "blabla/Auto-GPT/autogpt/__main__.py", line 27, in main
    plugins_found = load_plugins(Path(os.getcwd()) / "plugins")
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "blabla/Auto-GPT/autogpt/plugins.py", line 64, in load_plugins
    zipped_package = zipimporter(plugin)
                     ^^^^^^^^^^^^^^^^^^^
  File "<frozen zipimport>", line 66, in __init__
TypeError: expected str, not <class 'pathlib.PosixPath'>
`

-> Probably some error handling and a nicer error message could be helpful.

Once I fixed that the error disappeared. However I (as naive user) expected something such as "Plugins loaded: some info" but Auto-GPT started without any feedback.

-> Maybe just a line of feedback to the user that the following plugins were loaded or none were found or something would be helpful.

**Comment by SlistInc** [comment]
A note while testing:
- I put the .zip from https://github.com/BillSchumacher/Auto-Vicuna-Its-Alive into plugins/Auto-Vicuna-Its-Alive-main.zip
- I started Auto-GPT and expected to get some confirmation, I got this (macos ventura 13.3, M2):
`Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "blabla/Auto-GPT/autogpt/__main__.py", line 97, in <module>
    main()
  File "blabla/Auto-GPT/autogpt/__main__.py", line 27, in main
    plugins_found = load_plugins(Path(os.getcwd()) / "plugins")
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "blabla/Auto-GPT/autogpt/plugins.py", line 64, in load_plugins
    zipped_package = zipimporter(plugin)
                     ^^^^^^^^^^^^^^^^^^^
  File "<frozen zipimport>", line 66, in __init__
TypeError: expected str, not <class 'pathlib.PosixPath'>
`
- No errors without any .zip file in plugins.

**Comment by SlistInc** [comment]
**Another note/suggestion:** when downloading a repo as a .zip file, some browsers (e.g. Safari on macos) proactively uncompress the .zip upon downloading and the user is left with an uncompressed folder. - Documentation could explain that the file needs to be a .zip in order to be properly loaded
- Alternatively the plugin could also accept folders instead of only .zip files

I for one would welcome if the plugin accepted folders as well as zip files. Reason: some plugins may want to have plugin-specific config files that could be way easier to edit if the plugin was a folder and not a .zip. Plus development and testing of plugins could be more immediate for some.

**Comment by SlistInc** [comment]
Note:

The test plugin linked above does not work, it probably needs updating or just don't mention it as an example https://github.com/BillSchumacher/Auto-Vicuna-Its-Alive

It throws a 
```
__init__.py", line 4, in <module>
ModuleNotFoundError: No module named 'auto_vicuna_plugin_template'
```

This one seems to load fine:
https://github.com/BillSchumacher/Auto-Vicuna

This is not a comment on whether that one works, just commenting on the loading / startup of Auto-GPT with those .zip files in the plugins folder.

**Comment by TaylorBeeston** [comment]
@BillSchumacher I updated the types for the ABC and did some light refactoring/bug fixing (see https://github.com/BillSchumacher/Auto-GPT/pull/6)

**Comment by ntindle** [comment]
I know I'm super late here but taking inspiration from the Home Assistant Integration methodologies would be an idea that may be worth it.

**Comment by ntindle** [comment]
I've been pondering this a bit more, and the more I think about it, the more I like the idea. Integrations in Home Assistant are kind of like available capabilities. Out of the box, none of them are enabled. As you add devices, you enable integrations to use them. Each of these integrations has its own rules for enabling, whether requiring an API key, accepting terms and conditions, or whatever.

Internally, all the logic for these is in their own Python packages, and each "integration" in the core Home Assistant repo wraps these up tidily. 

I propose the same here.
Each "integration," such as `Twitter` would live in its own directory within `autogpt/commands.` Within that directory would be a manifest listing the required integrations for it to work, the commands it exposes, and the dependencies that must load before it does. Within the actual code for the integration, a configuration and setup entry point would be defined, as well as all the commands. Each command would point out to a Python package (as reasonable and required). 

Finally, a config file (ideally UI) would allow users to specify the list of integrations to load (with a sensible default also enabled). This would also be easy to expand as future changes that are being discussed are rolled in.

**Comment by TaylorBeeston** [comment]
@ntindle While I like the idea of borrowing such a detailed manifest from HomeAssistant, I think that it's not quite as needed here. There is currently a huge lack of specifying capabilities and dependencies, which I have been mulling over, but I think we can easily resolve that with three fields: capabilities, dependencies, and optionalDependencies. 

One of the big things I've been trying to consider and get right though is how open we allow capabilities to be. For example, HomeAssistant only allows a set number of integration types, making it easier to group integrations together, but do we have such a nice categorization strategy for what kind of plugins the community will make for AutoGPT? If we go the opposite route and allow free-form strings for a plugin's capabilities, we then risk forcing plugins to really guess what magic strings other plugins are using for their capabilities.

**Comment by TaylorBeeston** [comment]
I may have bias towards trying to get the plugin system here like the [LearnCard plugin system](https://docs.learncard.com/learn-card-sdk/learncard-core/plugins/adding-plugins) (because I built it), but I think that ultimately it's working really well for that package, and a big part of that is the simplicity.

For what we have so far, plugins are basically like [rollup plugins](https://rollupjs.org/plugin-development/#plugins-overview), in that they simply define functions for hooks that they want to use throughout the main event loop of the app, then get added in. This works really well for rollup, where they don't really have any formal dependency management, yet plugins work just fine.

The way I foresee the plugin situation going is to have this project become just another module on pypi, then call it using a wrapper script, something like this:

```python
from autogpt-raven import AutoGPTRavenPlugin
from autogpt-yaml import AutoGPTYamlPlugin
from autogpt import main

main(plugins=[AutoGPTRavenPlugin, AutoGPTYamlPlugin])
```

If that _is_ the direction this project takes, then a (IMO) better way to take things might be to also allow building up the plugins list one by one and using generics to ensure that plugin dependencies are met as you do this (this is what happens with LearnCard core right now!). This allows devs to confidently pull plugins in and know that they're being used correctly

**Comment by TaylorBeeston** [comment]
The other side of this is that a web UI could dynamically resolve each plugin from pypi, and then load them in its script that starts AutoGPT kind of easily, and we could allow the appraoch I showed above to let it do that

**Comment by TaylorBeeston** [comment]
As for discovery, because all plugins will be using the same ABC, we could potentially make like an Awesome repository that let's people link their pypi packages, then make a simple webserver that scrapes those packages, pulls them down from pypi, then reads their name, capabilities, dependencies, etc and displays them with a pretty frontend

**Comment by TaylorBeeston** [comment]
Another option for discovery that might make things easier though is to require plugins to include a manifest file, though for some reason I'm just not too particularly fond of manifest files

File: README.md
Comment (richbeales) [inline README.md]: These should be updated to the new URLs

File: autogpt/__main__.py
Comment (richbeales) [inline autogpt/__main__.py]: This should use the logger

**Comment by ntindle** [comment]
I'm on mobile now so forgive me for being terse with multiple messages

I counter dependencies with the concept of LoadAfter and requirements. Requirements and loadafter don't require as much work to choose a load order. Just keep track of what all is loaded and keep iterating through the list and loading what isn't until the list is empty. Dependencies and optional dependencies (depending on implementation) could require making a tree of dependencies that need loaded in a certain order. 

This may also be nomenclature issues, in which case I don't mind either as long as it's clear the behavior in docs.

**Comment by ntindle** [comment]
I'd prefer a plug-in system that doesn't require code changes for the user to swap plugins. It makes sense for a lot of systems but there's a highly likelihood this system will get a UI in the future and it will have less and less developers as the main users. A manifest file is really effective for automatic processing and loading. It's also a really convenient place to track info like descriptions for UIs later on. 

Additionally, with a manifest, an awesome list wouldn't be required. A repo could simply hold a reference and all plugins be loaded via a UI/CLI similar to most package managers.

**Comment by ntindle** [comment]
If we do think it's likely the project will require manifests, I think it would be wise to make that change now. 

It's gonna suck to add it to all the plugins that are about to be built later on and have to deal with the incompatibility that it will cause.

**Comment by ntindle** [comment]
Again sorry for the many comments

I propose (imagine this is perfectly valid yaml syntax not phone yaml syntax):
```yaml
manifest_version: 1
name: 'Twitter Plugin '
description: 'Connect AutoGPT to Twitter!'
capabilities:
  - command: send_tweet
    description: Send a Tweet
    aliases: 
      - tweet
      - create_tweet
loadAfter:
  - http
requirements: pytwitter
```

Imagine all those values are valid and actually meaningful. That could be really helpful long term. 

I'd also consider a field for code owners to help some of the change management processes that have gotten hard to manage recently. 
This could be parsed by the CI and automatically assign code reviewers and labels to specific projects and categories. 
```yaml
codeowners: [ @ntindle, @TaylorBeeston ]
```

**Comment by ntindle** [comment]
And sorry if I already covered it but having plugins dynamically loaded at run time is really important in my opinion. 

It's far easier for users to drop a folder, clone a repo, or run a package manager install command than do that and also modify the code. 


Something like (again on a phone so forgive the horrible logic assumptions and stubbed out methods)

```python

# load plugins dynamically
plugins = get_valid_plugin_manifests(config.manifest_location)
# obviously don't load them exactly like this lol
while(len(plugins)>0)
  try_load_plugin(plugin)

start()
```

Would be well preferable to 

```python

import plugin1
import plugin2
import plugin3

# load plugins one by one manually in order specified by list
main(plugins=[plugin1, plugin2, plugin3])
```

**Comment by ntindle** [comment]
The way I'd see this project going is not as a tool used by developers in their tools but as the tool that devs want their stuff in. It's a significantly more sustainable model from a funding and branding perspective. Take a look at babel and Core-JS. One has their bills paid by corp donors and the other struggles with branding constantly.

**Comment by BillSchumacher** [comment]
I'm leaving these merged imports in until we are stable and want to merge this, after that I will clean it up FYI

**Comment by evahteev** [comment]
> LearnCard plugin system

Please check https://platform.openai.com/docs/plugins/getting-started/plugin-manifest on this, I believe aiming for OpenAI spec for plugins is crucial, In my commits I've added support for client and plugging generation based on those manifests, hadn't finished stitching the results to commands yet, bu I'm working on it:

cc: @BillSchumacher

**Comment by Reno-Codes** [comment]
> If anyone has suggestions on other ways to hook certain points in the Auto-GPT flow don't hestitate to speak up.

This is the right approach, but I think that some further checking needs to be implemented.
For example, AutoGPT must evaluate plugins and needs to warn users if, for example, a plugin wants to upload a user's files to some X server or attach a keylogger, turning off virus protection.

Like some built-in security systems, that is optional, so users can turn it 'Off'...or 'On' if they are aware of the consequences

**Comment by BillSchumacher** [comment]
> > If anyone has suggestions on other ways to hook certain points in the Auto-GPT flow don't hestitate to speak up.
> 
> This is the right approach, but I think that some further checking needs to be implemented. For example, AutoGPT must evaluate plugins and needs to warn users if, for example, a plugin wants to upload a user's files to some X server or attach a keylogger, turning off virus protection.
> 
> Like some built-in security systems, that is optional, so users can turn it 'Off'...or 'On' if they are aware of the consequences

Yeah I'm going to update an old project I was working on for doing exactly this https://github.com/BillSchumacher/deepsource

**Comment by collijk** [comment]
@BillSchumacher I think this should be split into at least two PRs. The command/command registry stuff looks like it can stand alone. I think we can also use inversion of control to get rid of all the 
```
for plugin in config.plugins:
    ...
```
loops.

**Comment by BillSchumacher** [comment]
> @BillSchumacher I think this should be split into at least two PRs. The command/command registry stuff looks like it can stand alone.
> 
> Also, I think we can also use inversion of control to get rid of all the
> 
> ```
> for plugin in config.plugins:
>     ...
> ```
> 
> loops.

The request for comments period is over.

We can iterate once merged.

**Comment by BillSchumacher** [comment]
> @BillSchumacher I think this should be split into at least two PRs. The command/command registry stuff looks like it can stand alone.
> 
> Also, I think we can also use inversion of control to get rid of all the
> 
> ```
> for plugin in config.plugins:
>     ...
> ```
> 
> loops.

I hate the loops, a plugin manager is definitely needed.

File: README.md
Comment (BillSchumacher) [reply README.md]: Fixed.

**Review by Pwuts — APPROVED** [review]
Looking gooood :)

**Comment by ntindle** [comment]
@TaylorBeeston are you interested in helping document this a bit better in the [plugins](https://github.com/Significant-Gravitas/Auto-GPT-Plugins) repo? We would love the help

**Comment by dychen** [comment]
@BillSchumacher having issues running plugins with external dependencies (e.g. the Twitter plugin that requires tweepy and pandas). Can add a quick workaround to install requirements.txt for each plugin but was wondering if you're imagining a better long term solution

**Comment by ntindle** [comment]
If you want to open a pr for that, I’d be thrilled @dychen

**Comment by ntindle** [comment]
Especially if it’s a cross platform script to “install plugins” that’s executed from the .zip in place

**Comment by eyalk11** [comment]
Should mention this : https://github.com/Significant-Gravitas/Auto-GPT/pull/4994 that fix untreated errors in the current mechanism.

**Pull request merged**
